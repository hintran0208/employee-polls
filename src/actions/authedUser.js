export const LOGOUT_AUTHENTICATED_USER = "LOGOUT_AUTHENTICATED_USER";
export const SET_AUTHENTICATED_USER = "SET_AUTHENTICATED_USER";

export function setAuthenticatedUser(authedUser) {
  return {
    type: SET_AUTHENTICATED_USER,
    authedUser,
  };
}

export function logoutAuthedUser() {
  return {
    type: LOGOUT_AUTHENTICATED_USER,
  };
}

export function handleLogin(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();

    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );

    if (!!user) {
      return dispatch(setAuthenticatedUser(user));
    }
  };
}

export function handleLogout() {
  return (dispatch) => {
    return dispatch(logoutAuthedUser());
  };
}
