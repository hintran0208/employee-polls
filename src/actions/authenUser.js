const LOGOUT_AUTHENTICATED_USER = "LOGOUT_AUTHENTICATED_USER";
const SET_AUTHENTICATED_USER = "SET_AUTHENTICATED_USER";

function setAuthenticatedUser(userId) {
  return {
    type: SET_AUTHENTICATED_USER,
    userId,
  };
}

function logoutAuthenticatedUser() {
  return {
    type: LOGOUT_AUTHENTICATED_USER,
  };
}

export {
  LOGOUT_AUTHENTICATED_USER,
  SET_AUTHENTICATED_USER,
  setAuthenticatedUser,
  logoutAuthenticatedUser,
};
