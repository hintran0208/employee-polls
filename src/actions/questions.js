const ADD_NEW_QUESTION = "ADD_NEW_QUESTION";
const SAVE_ANSWER_VOTE = "SAVE_ANSWER_VOTE";
const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

const receiveQuestions = (questions) => {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
};

const saveAnswerVote = (author, qid, answer) => {
  return {
    type: SAVE_ANSWER_VOTE,
    author,
    qid,
    answer,
  };
};

const addQuestion = (question) => {
  return {
    type: ADD_NEW_QUESTION,
    question,
  };
};

export {
  ADD_NEW_QUESTION,
  RECEIVE_QUESTIONS,
  SAVE_ANSWER_VOTE,
  addQuestion,
  receiveQuestions,
  saveAnswerVote,
};
