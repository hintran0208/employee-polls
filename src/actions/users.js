export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_NEW_ANSWER_USER = "ADD_NEW_ANSWER_USER";
export const ADD_NEW_QUESTION_USER = "ADD_NEW_QUESTION_USER";

function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

function addNewAnswer(authedUser, qid, answer) {
  return {
    type: ADD_NEW_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}

function addNewQuestion({ author, id }) {
  return {
    type: ADD_NEW_QUESTION_USER,
    author,
    id,
  };
}

export { addNewQuestion, addNewAnswer, receiveUsers };
