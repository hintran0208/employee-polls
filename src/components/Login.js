import { useState } from "react";
import { connect } from "react-redux";
import { setAuthenticatedUser } from "../actions/authenUser";

const Login = ({ dispatch, authorizedUser }) => {
  const [username, setUsername] = useState("sarahedo");
  const [password, setPassword] = useState("password123");

  const handleUserLogin = (username, password) => {
    return (dispatch, getState) => {
      const { users } = getState();

      const loginUser = Object.values(users).find(
        (user) => user.id === username && user.password === password
      );

      if (!!loginUser) {
        return dispatch(setAuthenticatedUser(loginUser));
      }
    };
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleUserLogin(username, password));
    setUsername("");
    setPassword("");
  };

  return (
    <div className="flex min-h-full flex-1 flex-col justify-center px-6 py-12 lg:px-8">
      <div className="sm:mx-auto sm:w-full sm:max-w-sm">
        <img
          className="mx-auto w-auto"
          src="https://t3.ftcdn.net/jpg/02/90/83/56/360_F_290835671_jFAkOJOyejt4uZ0b7NvZtoNlD92L7cHh.jpg"
          alt="Your Company"
        />
        <h2 className="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900">
          Sign in to your account
        </h2>
      </div>

      <div className="mt-10 sm:mx-auto sm:w-full sm:max-w-sm">
        <form className="space-y-6" onSubmit={handleSubmit}>
          <div>
            <label htmlFor="username" className="block text-sm font-medium leading-6 text-gray-900">
              Username
            </label>
            <div className="mt-2">
              <input
                id="username"
                name="username"
                type="username"
                value={username}
                onChange={(e) => setUsername(e.target.value)}
                required
                className="p-4 block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                data-testid="username"
              />
            </div>
          </div>

          <div>
            <div className="flex items-center justify-between">
              <label
                htmlFor="password"
                className="block text-sm font-medium leading-6 text-gray-900"
              >
                Password
              </label>
            </div>
            <div className="mt-2">
              <input
                id="password"
                name="password"
                type="password"
                autoComplete="current-password"
                value={password}
                onChange={(e) => setPassword(e.target.value)}
                required
                className="p-4 block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                data-testid="password"
              />
            </div>
          </div>

          <div>
            <button
              type="submit"
              className="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
              data-testid="submit"
            >
              Sign in
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authorizedUser: authedUser !== null,
});

export default connect(mapStateToProps)(Login);
