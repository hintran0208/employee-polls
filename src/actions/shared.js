import { receiveUsers } from "./users";
import { receiveQuestions, saveAnswerVote, addQuestion } from "./questions";
import { fetchInitialData } from "../api/api";
import { saveQuestion, saveQuestionAnswer } from "../api/api";
import { addNewAnswer, addNewQuestion } from "./users";

function handleInitialData() {
  return (dispatch) => {
    return fetchInitialData()
      .then(({ users, questions }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      })
      .catch((error) => {
        console.log("Cannot handle init data", error);
      });
  };
}

function handleAddVote(questionId, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestionAnswer(authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addNewAnswer(authedUser.id, questionId, answer));
        dispatch(saveAnswerVote(authedUser.id, questionId, answer));
      })
      .catch((error) => {
        console.log("Cannot handle add data", error);
      });
  };
}

function handleAddNewQuestion(firstVote, secondVote) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestion(firstVote, secondVote, authedUser)
      .then((question) => {
        dispatch(addQuestion(question));
        dispatch(addNewQuestion(question));
      })
      .catch((error) => {
        console.log("Cannot handle add data", error);
      });
  };
}

export { handleInitialData, handleAddVote, handleAddNewQuestion };
