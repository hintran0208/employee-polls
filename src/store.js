import { configureStore } from "@reduxjs/toolkit";
import authedUser from "./reducers/authedUser";
import questions from "./reducers/questions";
import users from "./reducers/users";

export const store = configureStore({
  reducer: {
    authedUser,
    users,
    questions,
  },
});
