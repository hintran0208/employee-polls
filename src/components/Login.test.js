import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Login from "./Login";
import { handleInitialData } from "../actions/shared";

describe("Login", () => {
  it("should reset input value after submitted", async () => {
    await store.dispatch(handleInitialData());

    const utils = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const userNameEl = screen.getByTestId("username");
    const passWordEl = screen.getByTestId("password");
    const submitEl = screen.getByTestId("submit");

    fireEvent.change(userNameEl, { target: { value: "incorrectUsername" } });
    fireEvent.change(passWordEl, { target: { value: "incorrectPassword" } });

    expect(userNameEl.value).toBe("incorrectUsername");
    expect(passWordEl.value).toBe("incorrectPassword");

    fireEvent.click(submitEl);

    expect(userNameEl.value).toBe("");
    expect(passWordEl.value).toBe("");
  });
});
