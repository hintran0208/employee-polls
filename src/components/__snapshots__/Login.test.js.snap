// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the login component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex min-h-full flex-1 flex-col justify-center px-6 py-12 lg:px-8"
      >
        <div
          class="sm:mx-auto sm:w-full sm:max-w-sm"
        >
          <img
            alt="Your Company"
            class="mx-auto w-auto"
            src="https://t3.ftcdn.net/jpg/02/90/83/56/360_F_290835671_jFAkOJOyejt4uZ0b7NvZtoNlD92L7cHh.jpg"
          />
          <h2
            class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
          >
            Sign in to your account
          </h2>
        </div>
        <div
          class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
        >
          <form
            class="space-y-6"
          >
            <div>
              <label
                class="block text-sm font-medium leading-6 text-gray-900"
                for="username"
              >
                Username
              </label>
              <div
                class="mt-2"
              >
                <input
                  class="p-4 block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  id="username"
                  name="username"
                  required=""
                  type="username"
                  value="sarahedo"
                />
              </div>
            </div>
            <div>
              <div
                class="flex items-center justify-between"
              >
                <label
                  class="block text-sm font-medium leading-6 text-gray-900"
                  for="password"
                >
                  Password
                </label>
              </div>
              <div
                class="mt-2"
              >
                <input
                  autocomplete="current-password"
                  class="p-4 block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value="password123"
                />
              </div>
            </div>
            <div>
              <button
                class="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
                type="submit"
              >
                Sign in
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex min-h-full flex-1 flex-col justify-center px-6 py-12 lg:px-8"
    >
      <div
        class="sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <img
          alt="Your Company"
          class="mx-auto w-auto"
          src="https://t3.ftcdn.net/jpg/02/90/83/56/360_F_290835671_jFAkOJOyejt4uZ0b7NvZtoNlD92L7cHh.jpg"
        />
        <h2
          class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
        >
          Sign in to your account
        </h2>
      </div>
      <div
        class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <form
          class="space-y-6"
        >
          <div>
            <label
              class="block text-sm font-medium leading-6 text-gray-900"
              for="username"
            >
              Username
            </label>
            <div
              class="mt-2"
            >
              <input
                class="p-4 block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                id="username"
                name="username"
                required=""
                type="username"
                value="sarahedo"
              />
            </div>
          </div>
          <div>
            <div
              class="flex items-center justify-between"
            >
              <label
                class="block text-sm font-medium leading-6 text-gray-900"
                for="password"
              >
                Password
              </label>
            </div>
            <div
              class="mt-2"
            >
              <input
                autocomplete="current-password"
                class="p-4 block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                id="password"
                name="password"
                required=""
                type="password"
                value="password123"
              />
            </div>
          </div>
          <div>
            <button
              class="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
              type="submit"
            >
              Sign in
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
